lib/src/http/utils.dart

```diff
@@ -46,22 +48,77 @@ Duration calculateBackOff(RequestAutoRetry autoRetry) {
   return backOff;
 }
 
-Duration _calculateExponentialBackOff(RequestAutoRetry autoRetry) {
+Duration _calculateExponentialBackOff(RequestAutoRetry autoRetry,
+    {@visibleForTesting Random random}) {
+  if (autoRetry.backOff.withJitter ?? false) {
+    final jitteredBackOff = _calculateAdvancedExponentialJitteredBackOffInMs(
+        autoRetry,
+        random: random);
+    // If we're over the maximum duration, fall back to a fixed maxInterval with full jitter
+    if (jitteredBackOff > autoRetry.backOff.maxInterval.inMilliseconds) {
+      return Duration(
+          milliseconds:
+              (autoRetry.backOff.maxInterval.inMilliseconds.toDouble() *
+                      (random ?? Random()).nextDouble())
+                  .toInt());
+    }
+    return Duration(milliseconds: jitteredBackOff);
+  }
+
+  return Duration(
+      milliseconds: _calculateUnjitteredExponentialBackOffInMs(autoRetry));
+}
+
+int _calculateUnjitteredExponentialBackOffInMs(autoRetry) {
   int backOffInMs = autoRetry.backOff.interval.inMilliseconds *
       pow(_exponentialBase, autoRetry.numAttempts);
-  backOffInMs = min(autoRetry.backOff.maxInterval.inMilliseconds, backOffInMs);
+  return min(autoRetry.backOff.maxInterval.inMilliseconds, backOffInMs);
+}
 
-  if (autoRetry.backOff.withJitter == true) {
-    final random = Random();
-    backOffInMs = random.nextInt(backOffInMs);
+/// Returns the jittered backoff delay in ms using an advanced jittering algorithm.
+///
+/// Taken from https://github.com/Polly-Contrib/Polly.Contrib.WaitAndRetry/blob/master/src/Polly.Contrib.WaitAndRetry/Backoff.DecorrelatedJitterV2.cs#L35-L65
+/// See the details here: https://github.com/Polly-Contrib/Polly.Contrib.WaitAndRetry#wait-and-retry-with-jittered-back-off
+///
+/// See RFD 277 where this was originally proposed: https://sandbox.wdesk.com/a/QWNjb3VudB82NzE5NTMwMjcyOTQ4MjI0/doc/0c0c78a2bba448babb8bb2e45ba62f7b/r/-1/v/1/sec/0c0c78a2bba448babb8bb2e45ba62f7b_4
+int _calculateAdvancedExponentialJitteredBackOffInMs(RequestAutoRetry autoRetry,
+    {@visibleForTesting Random random}) {
+  // We subtract 1 from the numAttempts since the algorithm uses previous
+  // _retry_ attempts, and [tracker.numAttempts] is _total_ attempts, meaning
+  // it will always be 1 greater than the number of _retry_ attempts.
+  final t = autoRetry.numAttempts.toDouble() -
+      1.0 +
+      (random ?? Random()).nextDouble();
+  final next = pow(2, t) * _tanh(sqrt(4.0 * t));
+  final unscaledBackOff = next - autoRetry.previous;
+  final backoffInMs =
+      unscaledBackOff * 1 / 1.4 * (autoRetry.backOff.interval.inMilliseconds);
+  autoRetry.previous = next;
+  return backoffInMs.toInt();
+}
+
+/// Calculate the hyperbolic tangent of [angle] in radians.
+///
+/// Taken from `dart_numerics` package, which is not used here because it is not
+/// supported for web applications.
```

Comment:
We'll need to cite the source link and add their license to ours. I'll find an example of where we do that elsewhere

Reply:
https://github.com/Workiva/dart_codemod/blob/bc209e6a61f4b06ff6d9e7051b69b3307ecb77b1/lib/src/logging.dart#L15-L16

https://github.com/Workiva/dart_codemod/blob/master/LICENSE#L15-L30